### PR TITLE
Update layer list context keys

### PR DESCRIPTION
**Pull request by alisterburt**
# Description
This PR updates the context keys available on the `LayerList`, increasing the scope of what can be inferred from our `LayerList` context

- missing `only_<layer_class>_selected` context keys added
- `num_<layer_class>_selected` context keys added
- updated some keys to unify naming style

Internal usages of layer list context keys have been updated to reflect the new names. @tlambert03 assures me that these shouldn't be being used by others yet and changing the names to unify is okay at this stage.

@napari/core-devs could you take a look at these `ContextKey` names? Once these get out in the open any changes will be hard to revert 👀 

## Type of change
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# References
@tlambert03 + https://napari.org/guides/contexts_expressions.html

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

**Comment by tlambert03** [comment]
to make this a little easier to review/discuss, here is the new complete list of layerlist context keys.  The primary questions we should be asking ourselves are:

1. Could we express all of the conditions about the current state of the layerlist that we might want to with these keys? 
2. Is this a minimal set? (i.e. could we capture a similar expressiveness with less keys)
3. Are the key names themselves exaclty how we want them?  These will be used in npe2 manifest to express "when" conditions... e.g. `num_selected_layers == 3 and num_selected_labels_layers > 1`

(reminder: we can't use callables here ... everything needs to be expressible as an [expression of values and operations](https://napari.org/guides/contexts_expressions.html?highlight=context#napari-expressions), not function calls)

| key                          | description                                                                         |
|------------------------------|-------------------------------------------------------------------------------------|
| num_selected_layers          | Number of currently selected layers.                                                |
| all_layers_linked            | True when all selected layers are linked.                                           |
| num_unselected_linked_layers | Number of unselected layers linked to selected layer(s).                            |
| active_layer_is_rgb          | True when the active layer is RGB.                                                  |
| active_layer_type            | Lowercase name of active layer type, or None of none active.                        |
| only_image_layers_selected   | True when there is at least one selected layer and all selected layers are images   |
| num_selected_image_layers    | Number of selected image layers.                                                    |
| only_labels_layers_selected  | True when there is at least one selected layer and all selected layers are labels   |
| num_selected_labels_layers   | Number of selected labels layers.                                                   |
| only_points_layers_selected  | True when there is at least one selected layer and all selected layers are points   |
| num_selected_points_layers   | Number of selected points layers.                                                   |
| only_shapes_layers_selected  | True when there is at least one selected layer and all selected layers are shapes   |
| num_selected_shapes_layers   | Number of selected shapes layers.                                                   |
| only_surface_layers_selected | True when there is at least one selected layer and all selected layers are surfaces |
| num_selected_surface_layers  | Number of selected surface layers.                                                  |
| only_vectors_layers_selected | True when there is at least one selected layer and all selected layers are vectors  |
| num_selected_vectors_layers  | Number of selected vectors layers.                                                  |
| only_tracks_layers_selected  | True when there is at least one selected layer and all selected layers are tracks   |
| num_selected_tracks_layers   | Number of selected tracks layers.                                                   |
| active_layer_ndim            | Number of dimensions in the active layer, or `None` if nothing is active.           |
| active_layer_shape           | Shape of the active layer, or `None` if nothing is active.                          |
| active_layer_dtype           | Dtype of the active layer, or `None` if nothing is active.                          |
| all_layers_same_shape        | True when all selected layers have the same shape.                                  |

for me, I think it looks reasonable... the keys for every layer type are a bit ugly, but I can't immediately think of alternatives 
only things I found myself wondering:

- do we like the longer `only_vectors_layers_selected` over `only_vectors_selected`?
- do we need any of the `only_<type>_layers_selected` anymore?  can't that be expressed as `num_selected_layers == num_selected_<type>_layers`?

**Comment by alisterburt** [comment]
thanks for the feedback/structure for discussion @tlambert03 

> do we like the longer only_vectors_layers_selected over only_vectors_selected?

Overall I prefer the slightly longer key but don't feel strongly. 

I can imagine times we would want to have context around the number of points/shapes selected -> because some of these names might clash (num_points_selected vs. num_points_layers_selected) I thought it would be useful to be explicit. These keys would likely live in a different context (i.e. not the layerlist so this may be a non-issue)

> do we need any of the only_<type>_layers_selected anymore? can't that be expressed as num_selected_layers == num_selected_<type>_layers?

It definitely can, I've gone back and forth on this but I now think that starting out with the minimal subset of keys is probably best - will update :)

**Comment by DragaDoncila** [comment]
I have only semi closely followed this discussion but I do want to say that:

- I definitely agree with keeping the longer names for the sake of explicitness and to keep the door open for context keys about the layer contents
- I'm OK with dropping the `only_x_layers_selected` keys, but the equivalent statement should be well documented for users, as should any other common "derived keys" that we can think of

**Comment by sofroniewn** [comment]
Amazing to see this kickoff. I see @alisterburt and @DragaDoncila have both weighed in during the time it has taken me to write this - it looks like we're all in agreement here

> do we need any of the only_<type>_layers_selected anymore? can't that be expressed as num_selected_layers == num_selected_<type>_layers?

I think we can start without them, always easy to add convenience keys later, but focusing on the minimal set at the beginning I think is good.

> These will be used in npe2 manifest to express "when" conditions... e.g. num_selected_layers == 3 and num_selected_labels_layers > 1

This is great! I'm guessing that we can use the same validation approach to "when" like we've done with locations in https://github.com/napari/npe2/pull/161 and that this information will get added to npe2

> Could we express all of the conditions about the current state of the layerlist that we might want to with these keys?

I think we should add **num_layers**.

I'd rename `all_layers_same_shape` to `all_selected_layers_same_shape`. I know it's long, but it is more clear.
I think we should add **all_selected_layers_same_type**, might be important to use alongside `all_selected_layers_same_shape`. Although in theory it is computable from other keys it would be a very long OR statement!

I was thinking we could replace ~~all_layers_linked~~ with **num_selected_linked_layers** as you could then compare with "num_selected_layers" to get "all", but you could actually have two different linking groups within the selected group, so it could be a bit harder. The linking context is the only part that might not still quiet perfect, but I think is good enough to get going

If we accept all the changes I propose then the table becomes: (note i have rearranged some order too to group things more)

| key                          | description                                                                         |
|------------------------------|-------------------------------------------------------------------------------------|
| **num_layers**         | Number of currently selected layers.                                                |
| num_selected_layers          | Number of currently selected layers.                                                |
| num_selected_image_layers    | Number of selected image layers.                                                    |
| num_selected_labels_layers   | Number of selected labels layers.                                                   |
| num_selected_points_layers   | Number of selected points layers.                                                   |
| num_selected_shapes_layers   | Number of selected shapes layers.                                                   |
| num_selected_surface_layers  | Number of selected surface layers.                                                  |
| num_selected_vectors_layers  | Number of selected vectors layers.                                                  |
| num_selected_tracks_layers   | Number of selected tracks layers.                                                   |
| **all_selected_layers_same_type**       | True when all selected layers have the same type.                                  |
| all_**selected**_layers_same_shape        | True when all selected layers have the same shape.                                  |
| all_**selected**_layers_linked        | True when all selected layers are linked.                                  |
| num_unselected_linked_layers | Number of unselected layers linked to selected layer(s).                            |
| active_layer_is_rgb          | True when the active layer is RGB.                                                  |
| active_layer_type            | Lowercase name of active layer type, or None of none active.                        |
| active_layer_ndim            | Number of dimensions in the active layer, or `None` if nothing is active.           |
| active_layer_shape           | Shape of the active layer, or `None` if nothing is active.                          |
| active_layer_dtype           | Dtype of the active layer, or `None` if nothing is active.                          |

**Review by andy-sweet — APPROVED** [review]
I like the PR as is, but like the summarized table from @sofroniewn even more - i.e. the smaller set of context keys with the `only*` keys.

**Comment by alisterburt** [comment]
waiting for CI to run but I think this is good to go now, I've incorporated nearly all changes suggested by @sofroniewn, in usage `all_selected_layers_same_type & num_selected_*_layers >= 1` feels nice as a clean proxy for `only_*_layers_selected`!

The only thing missing from the table @sofroniewn outlines is `num_layers` which is due to napari/napari#4529 - there is a path forwards outlined there but it's a larger change that will bloat this PR, I think we should press on with what we have here and circle back to having a reference to the `LayerList` on the selection 🙂

**Review by tlambert03 — APPROVED** [review]

**Comment by tlambert03** [comment]
looks like a couple real errors in typing?

```
napari/utils/context/_layerlist_context.py:121: error: Incompatible types in assignment (expression has type "List[Layer]", variable has type "Selection[Layer]")  [assignment]
napari/utils/context/_layerlist_context.py:125: error: Value of type "Selection[Layer]" is not indexable  [index]
```

just assign to a new variable (`_s = list(s)`) ...

File: napari/utils/context/_layerlist_context.py
Comment (tlambert03) [inline napari/utils/context/_layerlist_context.py]: what about?

```suggestion
    return len({x._type_string for x in s}) == 1
```

Comment (alisterburt) [reply napari/utils/context/_layerlist_context.py]: 🤣 I love it - thanks!!

**Comment by alisterburt** [comment]
@tlambert03 your suggestion should fix those typing errors too, cheers!

**Review by Czaki — DISMISSED** [review]
I do no see update of [guides/context_expressions.md](https://github.com/napari/napari/blob/e39b79bea1671b9a265e9c55316415da0ede1e19/docs/guides/contexts_expressions.md#napari-context-keys) in this PR, but this file contains table of example context keys that will be outdated after this PR merge. Could you update it? 

Is there place where all context keys are documented? or only code and PR comments (I'm not familiar with this part of napari code)?

**Review by sofroniewn — APPROVED** [review]

**Comment by tlambert03** [comment]
> I do no see update of [guides/context_expressions.md](https://github.com/napari/napari/blob/e39b79bea1671b9a265e9c55316415da0ede1e19/docs/guides/contexts_expressions.md#napari-context-keys) in this PR, but this file contains table of example context keys that will be outdated after this PR merge. Could you update it? 
> 
> 
> 
> Is there place where all context keys are documented? or only code and PR comments (I'm not familiar with this part of napari code)? 
> 
> 
> 
> 


That doc can be considered internal for napari core devs (as stated at the top).  I do have a plan for (auto)generating front facing docs using the descriptions jn these ContextKey objects... But until a plugin can actually take advantage of these keys using `when` conditions, we're not going to break anything by changing these here

**Review by jni — APPROVED** [review]

**Comment by Carreau** [comment]
@Czaki I've opened an issue mentioning that docs need to be updated, that way I can dismiss your review here (it's still tracked in napari/docs#146) and merge here, so that we move forward with some of the PRs and have less in flight.

**Comment by tlambert03** [comment]
@Czaki, in general, if you hit the request changes button, please note that it's a merge blocker.  If someone directly answers your question, please either remove your blocker, or ask for further clarification.  We'd prefer not to have to dismiss your reviews ourself, but if left there, someone needs to make a judgement call.  (@sofroniewn [had to make a similar dismissal](https://github.com/napari/napari/pull/4519#event-6679316786) last week)

**Comment by Czaki** [comment]
Ok. I do not know that. I will accept after my remarks were addressed (maybe not as first if I fully do not understand).

**Comment by lucyleeow** [comment]
I know this is old, but just a question, I notice in this PR the `only_*` functions have been replaced (e.g., `only_points`) but the old `only_*` function remains defined but not used later `ContextKey`. Is this the intention? Why not remove them since they are currently not used, or create a `ContextKey` for them if we think they may be useful later?

I could not find an answer in the comments.

**Pull request merged**
